core/engine/src/object/jsobject.rs

```diff
@@ -188,25 +221,9 @@ impl JsObject {
     ///
     /// Panics if the object is currently mutably borrowed.
     pub fn downcast<T: NativeObject>(self) -> Result<JsObject<T>, Self> {
-        if self.borrow().is::<T>() {
-            let ptr: NonNull<GcBox<VTableObject<dyn NativeObject>>> = Gc::into_raw(self.inner);
-
-            // SAFETY: the rooted `Gc` ensures we can read the inner `GcBox` in a sound way.
-            #[cfg(debug_assertions)]
-            unsafe {
-                let erased = ptr.as_ref();
-
-                // Some sanity checks to ensure we're doing the correct cast.
-                assert_eq!(size_of_val(erased), size_of::<GcBox<VTableObject<T>>>());
-                assert_eq!(align_of_val(erased), align_of::<GcBox<VTableObject<T>>>());
-            }
-
-            let ptr: NonNull<GcBox<VTableObject<T>>> = ptr.cast();
-
-            // SAFETY: The conversion between an `Any` and its downcasted type must be valid.
-            // The pointer returned by `Gc::into_raw` is the same one that is passed to `Gc::from_raw`,
-            // just downcasted to the type `T`.
-            let inner = unsafe { Gc::from_raw(ptr) };
+        if self.is::<T>() {
+            // SAFETY: We have verified that the object is of type `T`, so we can safely cast it.
+            let inner = unsafe { Gc::cast_unchecked::<VTableObject<T>>(self.inner) };
```

Comment:
Consider using the `JsObject` downcast method :)

```suggestion
            let inner = unsafe { self.downcast_unchecked::<T>() };
```